timely/src/dataflow/operators/generic/handles.rs

```diff
@@ -61,35 +59,43 @@ impl<'a, T: Timestamp, D: Container, P: Pull<BundleCore<T, D>>> InputHandleCore<
         })
     }
 
-    /// Repeatedly calls `logic` till exhaustion of the available input data.
-    /// `logic` receives a capability and an input buffer.
-    ///
-    /// # Examples
-    /// ```
-    /// use timely::dataflow::operators::ToStream;
-    /// use timely::dataflow::operators::generic::Operator;
-    /// use timely::dataflow::channels::pact::Pipeline;
-    ///
-    /// timely::example(|scope| {
-    ///     (0..10).to_stream(scope)
-    ///            .unary(Pipeline, "example", |_cap, _info| |input, output| {
-    ///                input.for_each(|cap, data| {
-    ///                    output.session(&cap).give_vec(&mut data.replace(Vec::new()));
-    ///                });
-    ///            });
-    /// });
-    /// ```
+    /// Reads the next input buffer (at some timestamp `t`) and a corresponding capability for `t`.
+    /// The timestamp `t` of the input buffer can be retrieved by invoking `.time()` on the capability.
+    /// Returns `None` when there's no more data available.
     #[inline]
-    pub fn for_each<F: FnMut(CapabilityRef<T>, RefOrMut<D>)>(&mut self, mut logic: F) {
-        let mut logging = self.logging.take();
-        while let Some((cap, data)) = self.next() {
-            logging.as_mut().map(|l| l.log(crate::logging::GuardedMessageEvent { is_start: true }));
```

Comment:
We can make the `next` and `next_mut` methods do this logging if desired. It wasn't clear to me if the current thinking was more in line with `next` that didn't log or with `for_each` that did log. Let me know if I got this the other way around